node/src/common/JXCoreEngine.cpp

```diff
@@ -0,0 +1,456 @@
+            }
+
+            JX_Free(reinterpret_cast<JXValue*>(_callScriptFunction));
+            delete reinterpret_cast<JXValue*>(_callScriptFunction);
+            _callScriptFunction = nullptr;
+
+            JX_StopEngine();
+            _started = false;
+        }
+        catch (...)
+        {
+            LogError("Failed to stop JXCore engine.");
+            callback(std::current_exception());
+            return;
+        }
+
+        LogVerbose("Stopped JXCore engine.");
+        callback(nullptr);
+    });
+}
+
+void JXCoreEngine::CallScript(
+    std::string scriptCode,
+    std::function<void(std::string resultJson, std::exception_ptr ex)> callback)
+{
+    LogTrace("JXCoreEngine::CallScript(\"%s\")", scriptCode.c_str());
+
+    _dispatcher.Dispatch([this, scriptCode, callback]()
+    {
+        this->CallScriptInternal(scriptCode, callback);
+    });
+}
+
+void JXCoreEngine::RegisterCallFromScript(
+    std::string scriptFunctionName,
+    std::function<void(std::string argsJson)> callback)
+{
+    LogTrace("JXCoreEngine::RegisterCallFromScript(\"%s\")", scriptFunctionName.c_str());
+
+    _dispatcher.Dispatch([this, scriptFunctionName, callback]()
+    {
+        if (!_started)
+        {
+            _initialCallFromScriptMap.emplace(scriptFunctionName, callback);
+        }
+        else
+        {
+            this->RegisterCallFromScriptInternal(scriptFunctionName, callback);
+        }
+    });
+}
+
+void JXCoreEngine::CallScriptInternal(
+    std::string scriptCode,
+    std::function<void(std::string resultJson, std::exception_ptr ex)> callback)
+{
+    try
+    {
+        if (!_started)
+        {
+            LogErrorAndThrow("JXCore engine is not started.");
+        }
+
+        std::function<void(std::string, std::exception_ptr)>* callbackPtr =
+        new std::function<void(std::string, std::exception_ptr)>(callback);
+
+        // The callback function pointer is passed through JavaScript as a hex-formatted number.
+        unsigned long long callId = reinterpret_cast<unsigned long long>(callbackPtr);
+        char callIdBuf[20];
+        snprintf(callIdBuf, sizeof(callIdBuf), "%llx", callId);
+
+        // Create JXValue arguments to the call-script function: callback pointer and script code string.
+        JXValue args[2];
+        JX_New(&args[0]);
+        JX_New(&args[1]);
+        JX_SetString(&args[0], callIdBuf);
+        JX_SetString(&args[1], scriptCode.c_str(), static_cast<int>(scriptCode.size()));
+
+        // Invoke the script function that will evaluate the provided script code then callback
+        // via the result or error callback.
+        JXValue unusedResult;
+        if (JX_CallFunction(reinterpret_cast<JXValue*>(_callScriptFunction), args, 2, &unusedResult))
+        {
+            JX_Free(&unusedResult);
+            LogVerbose("Successfully evaluated script code.");
+        }
+        else
+        {
+            LogErrorAndThrow("Failed to evaluate script code.");
```

Comment:
Yes, fixing.

---
In reply to: [71087053](https://github.com/openT2T/client/pull/1#discussion_r71087053) [](ancestors = 71087053)